js/ui/components/clubhouse.js

```diff
@@ -148,6 +153,111 @@ var ClubhouseWindowTracker = new Lang.Class({
     },
 });
 
+var getClubhouseAnimator = (function () {
+    let clubhouseAnimator;
+    return function() {
+        if (!clubhouseAnimator)
+            clubhouseAnimator = new ClubhouseAnimator();
+        return clubhouseAnimator;
+    };
+}());
+
+var ClubhouseAnimator = new Lang.Class({
+    Name: 'ClubhouseAnimator',
+    Extends: GObject.Object,
+
+    Signals: {
+        'animation-ready': {},
+    },
+
+    _init: function() {
+        this.parent();
+        this._proxy = null;
+        this._animations = {};
+        this._clubhousePaths = this._getClubhousePaths();
+    },
+
+    setClubhouseProxy: function(proxy) {
+        this._proxy = proxy;
+    },
+
+    _getClubhousePaths: function() {
+        let paths = [];
+        let installations = Flatpak.get_system_installations(null);
+        installations.unshift(Flatpak.Installation.new_user(null));
+
+        for (let installation of installations) {
+            let app = null;
+            try {
+                app = installation.get_current_installed_app(CLUBHOUSE_ID, null);
+            } catch (err if err.matches(Flatpak.Error, Flatpak.Error.NOT_INSTALLED)) {
+                continue;
+            }
+
+            if (app) {
+                let deploy_dir = app.get_deploy_dir();
+                paths.push(this._getActivateDir(deploy_dir));
+            }
+        }
+
+        return paths;
+    },
+
+    _getActivateDir: function(deployDir) {
+        // Replace the hash part of the deploy directory by "active", so the directory
+        // is always the most recent one (i.e. allows us to update the Clubhouse and
+        // still have the right dir).
+        let dir = deployDir.substr(-1) == '/' ? deployDir.slice(0, -1) : deployDir;
+        let splitDir = dir.split('/')
+
+        splitDir[splitDir.length - 1] = 'active';
+
+        return splitDir.join('/');
+    },
+
+    _getClubhouseUri: function(uri) {
+        // Discard the /app/ prefix
+        let uriSuffix = uri.replace(/^\/app\//g, '');
+
+        for (let path of this._clubhousePaths) {
+            let completeUri = path + '/files/' + uriSuffix;
+            if (GLib.file_test(completeUri, GLib.FileTest.EXISTS))
+                return completeUri;
+        }
+
+        return null;
+    },
+
+    getAnimationByUri: function(uri) {
+        let res = this._proxy.getAnimationMetadataRemote(uri, (metadata_str) => {
+            let json_data = null;
+
+            try {
+                json_data = JSON.parse(metadata_str);
+            } catch (e) {
+                log('Error parsing animation metadata json: ' + e);
+                return;
+            }
+
+            this._animations[uri] = json_data;
+            this.emit('animation-ready');
+        });
+    },
+
+    getAnimation: function(uri) {
```

Comment:
Expanding from my comment above, I'd like this object to only have one API entry point; it's confusing that `getAnimationByUri()` is an async method that ends up emitting a signal, but `getAnimation()` returns the `Animation` object that is ultimately what's needed.

Reply:
Please see my doubt in the reply to your previous comment. I will rename the `getAnimationByUri` to `loadAnimationByUri` in any case and this should avoid some of the confusion.

Reply:
I ended up implementing my suggestion and the code becomes indeed much simpler as I imagined.